### PR TITLE
Add JWT tests

### LINKED ISSUE #136: Add JWT test

### PR DESCRIPTION
**Pull request by funkyboy**
Fix #136 

- [X] Remove old test
- [X] Add tests that use `authURL`
- [X] Add tests that use `ClientOptions`
- [x] Add tests that use `authCallback`
- [x] Add tests for automatic renewal
- [x] Add tests with JWT wrapping an Ably native token `x-ably-token`
- [x] Add tests with encrypted token
- [x] Add tests with JWT that includes TokenDetails in `x-ably-capability`
- [x] Add example when token is returned with `application/jwt` Content-Type
- [x] Add refs to spec items
- [x] Change URL of `auth_url` when echo server is deployed. See TODOs

### CODE REVIEW

File: spec/acceptance/rest/auth_spec.rb [hunk 1]
```diff
@@ -1365,5 +1365,18 @@ def level=(new_level)
         expect(response).to be_a(Ably::Models::TokenDetails)
       end
     end
+
+    context 'when using JWT' do
+      it 'client pulls non nil stats when authenticating with a JWT encoded token' do
+        require 'jwt'
```
Comment (mattheworiordan) [inline spec/acceptance/rest/auth_spec.rb]: It's quite unusual to require this within a block.  I know it works, but it's not best practice.  Can we include all dependencies at the start of this file please?

**Comment by funkyboy** [comment]
Rebased on master now.

File: spec/acceptance/realtime/auth_spec.rb [hunk 2]
```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
...
+            end
+          end
+        end
+      end
+
+      context 'when the JWT token request includes a client_id' do
+        let(:client_id) { random_str }
+        let(:auth_callback) do
+          lambda do |token_params|
+            Ably::Rest::Client.new(default_options).auth.request_token({ client_id: client_id }, { auth_url: auth_url, auth_params: auth_params }).token
```
Comment (mattheworiordan) [inline spec/acceptance/realtime/auth_spec.rb]: This feels wrong to me, but I stand to be corrected about what is happening here.  Given the `auth_url` is getting a JWT token, and that JWT token has a `client_id` embedded, then why we are bothering to a) pass in a client_id into the request `TokenParams` given the `auth_url` in reality would be responsible for that, b) why even use the REST library here, we are testing the realtime library here i.e. isn't it clearer to show explicitly what is returned from `auth_callback` whereas right now it's just not clear if that is an Ably token, a JWT token, etc.

File: spec/acceptance/realtime/auth_spec.rb [hunk 3]
```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
...
+          end
+        end
+        let(:client_options) { default_options.merge(auto_connect: false, auth_callback: auth_callback) }
```
Comment (mattheworiordan) [inline spec/acceptance/realtime/auth_spec.rb]: Why auto_connect false?

File: spec/acceptance/realtime/auth_spec.rb [hunk 4]
```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
...
+          client.connect
+        end
+      end
+
+      context 'when the JWT token request includes a subscribe-only capability' do
+        let(:client_id) { random_str }
+        let(:basic_capability) { JSON.dump(channel_name => ['subscribe']) }
+        let(:auth_callback) do
+          lambda do |token_params|
+            Ably::Rest::Client.new(default_options).auth.request_token({ capability: basic_capability }, { auth_url: auth_url, auth_params: auth_params }).token
```
Comment (mattheworiordan) [inline spec/acceptance/realtime/auth_spec.rb]: As above, whilst I think a test for Rest clients generating tokens from `auth_urls` with `JWT`s is useful, I don't think we should be creating a level of indirection here that makes understanding this test that much hardder.  Why not simply make an HTTP request to the echo server directly and return that in `auth_callback`?

File: spec/acceptance/realtime/auth_spec.rb [hunk 5]
```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
...
+          client.connection.once(:connected) do
+            expect(client.auth.client_id).to eql(client_id)
+            stop_reactor
+          end
+          client.connect
+        end
+      end
+
+      context 'when the JWT token request includes a subscribe-only capability' do
+        let(:client_id) { random_str }
```
Comment (mattheworiordan) [inline spec/acceptance/realtime/auth_spec.rb]: This is unused

File: spec/acceptance/realtime/auth_spec.rb [hunk 6]
```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
...
+        let(:basic_capability) { JSON.dump(channel_name => ['subscribe']) }
+        let(:auth_callback) do
+          lambda do |token_params|
+            Ably::Rest::Client.new(default_options).auth.request_token({ capability: basic_capability }, { auth_url: auth_url, auth_params: auth_params }).token
+          end
+        end
+        let(:client_options) { default_options.merge(auth_callback: auth_callback) }
+
+        it 'client fails to publish to a channel with subscribe-only capability' do
+          channel = client.channels.get(channel_name)
```
Comment (mattheworiordan) [inline spec/acceptance/realtime/auth_spec.rb]: Should we not check that the connection is established, and that the channel the user does have permissions on is allowed?

File: spec/acceptance/rest/auth_spec.rb [hunk 7]
```diff
@@ -1338,5 +1338,31 @@ def coerce_if_time_value(field_name, value, params = {})
         expect(response).to be_a(Ably::Models::TokenDetails)
       end
     end
+
+    context 'when using JWT' do
+      let(:auth_url) { 'https://shrouded-plains-50367.herokuapp.com/createJWT' } # TODO: change this
+      let(:token) { Faraday.get "#{auth_url}?keyName=#{key_name}&keySecret=#{key_secret}" }
```
Comment (mattheworiordan) [inline spec/acceptance/rest/auth_spec.rb]: This is returning an HTTP response object, not a token

Comment (funkyboy) [reply spec/acceptance/rest/auth_spec.rb]: Same as above

File: spec/acceptance/realtime/auth_spec.rb [hunk 8]
```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
...
+      let(:auth_params) { { keyName: key_name, keySecret: key_secret } }
+      let(:channel_name) { 'test_JWT' }
+      let(:message_name) { 'message_JWT' }
+
+      context 'when using auth_url' do
+        let(:client_options) { default_options.merge(auto_connect: false, auth_url: auth_url, auth_params: auth_params) }
+
+        context 'when credentials are valid' do
+          it 'client successfully fetches a channel and publishes a message' do
+            message = client.channels.get(channel_name).publish message_name
```
[the remark this answers is not in the document]
Comment (funkyboy) [reply spec/acceptance/realtime/auth_spec.rb]: done https://github.com/ably/ably-ruby/pull/137/commits/8c9247798a46734e64f719502ad999ba0663180d

File: spec/acceptance/realtime/auth_spec.rb [hunk 9]
```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
...
+            client.connect
+          end
+        end
+
+        context 'when token is expired' do
+          let(:token_duration) { 5 }
+          let(:auth_params) { { keyName: key_name, keySecret: key_secret, expiresIn: token_duration } }
+          it 'receives a 40142 error from the server' do
+            client.connection.once(:connected) do
+              sleep token_duration + 1
```
[the remark this answers is not in the document]
Comment (funkyboy) [reply spec/acceptance/realtime/auth_spec.rb]: done https://github.com/ably/ably-ruby/pull/137/commits/57790ec84b3b01e022f938c4d49dfbb4ec80201e 
Sleep statement was a leftover. With a ttl of 5 we are beyond the limit to receive an AUTH message.

File: spec/acceptance/realtime/auth_spec.rb [hunk 10]
```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
...
+        let(:token_callback) do
+          lambda do |token_params|
+            Ably::Rest::Client.new(default_options).auth.request_token({}, { auth_url: auth_url, auth_params: auth_params, auth_method: :post}).token
+          end
+        end
+        let(:client_options) { default_options.merge(auth_callback: token_callback) }
+
+        context 'when credentials are valid' do
+          it 'authentication succeeds and stats are pulled correctly' do
+            client.stats do |stats|
```
[the remark this answers is not in the document]
Comment (funkyboy) [reply spec/acceptance/realtime/auth_spec.rb]: done https://github.com/ably/ably-ruby/pull/137/commits/93ee0fd265ac191d96b7de87793c30d595a83f14#diff-234b86ab2ae4f25eac36170940dfb2b1R1096

File: spec/acceptance/realtime/auth_spec.rb (hunk 5 above)
Comment (funkyboy) [reply spec/acceptance/realtime/auth_spec.rb]: done https://github.com/ably/ably-ruby/pull/137/commits/8baba6313a74a896ab68ee4b88995afe685b37e9

File: spec/acceptance/realtime/auth_spec.rb (hunk 3 above)
Comment (funkyboy) [reply spec/acceptance/realtime/auth_spec.rb]: Removed https://github.com/ably/ably-ruby/pull/137/commits/310c5a92b5b535df5afb86d7c0a5d0ba12de0ca6

File: spec/acceptance/realtime/auth_spec.rb [hunk 11]
```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
...
+            end
+            client.connect
+          end
+        end
+      end
+
+      context 'when using auth_callback' do
+        let(:token_callback) do
+          lambda do |token_params|
+            Ably::Rest::Client.new(default_options).auth.request_token({}, { auth_url: auth_url, auth_params: auth_params, auth_method: :post}).token
```
[the remark this answers is not in the document]
Comment (funkyboy) [reply spec/acceptance/realtime/auth_spec.rb]: I tried this but no luck :(
Do you have any idea as to why? https://github.com/ably/ably-ruby/pull/137/commits/93ee0fd265ac191d96b7de87793c30d595a83f14#diff-234b86ab2ae4f25eac36170940dfb2b1R1102

**Comment by funkyboy** [comment]
@paddybyers 

> generally I would have had the rest tests cover all of the cases that apply to REST, and then in realtime do the cases that additionally apply (which are checking the clientId in the CONNECTED response, and exercising the in-protocol reauth mechanism;

I think it's probably too late for this PR, but I'll take this into account for other repos.

> I don't know if I've missed it, but my reading of the auth-callback test here is that you're generating a regular Ably token, not a JWT;

The `/createJWT` only returns JWT tokens. Anyway I clarified/simplified some callbacks to make them more explicit: https://github.com/ably/ably-ruby/pull/137/commits/1ef2b12a240a642548acf847d3fedbc077583fdc and https://github.com/ably/ably-ruby/pull/137/commits/c1a56af82548a9669cc8592bbaaea1a4b04df816

> I don't see any code change or test for the application/jwt content type in an authUrl response - in order to get the echoserver to respond with each possible Content-Type, is another parameter required to be supported?

Added a new TODO about this in this PR and the echo server

File: spec/acceptance/realtime/auth_spec.rb [hunk 12]
```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
...
+            client.connect
+          end
+        end
+      end
+
+      context 'when the client is initialized with ClientOptions and the token is a JWT token' do
+        let(:client_options) { { token: token.body, environment: environment, protocol: protocol } }
+
+        context 'when credentials are valid' do
+          let(:token) { Faraday.get "#{auth_url}?keyName=#{key_name}&keySecret=#{key_secret}" }
```
[the remark this answers is not in the document]
Comment (mattheworiordan) [reply spec/acceptance/realtime/auth_spec.rb]: How is that possible @funkyboy?  Think about the following equivalent pseudo-code:

```ruby
token = Foo.bar
token.body == Foo.bar.body
```

Additionally:

```
Faraday.get("https://shrouded-plains-50367.herokuapp.com/createJWT").body
=> "Internal Server Error\n"
```

The problem is simply you are trying to get the body of the body in subsequent code, which is the problem this suggested change was trying to solve i.e. `token` is a string, not a response object.

File: spec/acceptance/realtime/auth_spec.rb [hunk 13]
```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
...
+          end
+        end
+        let(:client_options) { default_options.merge(use_token_auth: true, auth_callback: token_callback) }
+
+        it 'a new token is requested via auth_callback and the client gets reconnected' do
+          client.connection.once(:connected) do
+            original_token = auth.current_token_details
+            original_conn_id = client.connection.id
+
+            client.connection.once(:disconnected) do |state_change|
```
[the remark this answers is not in the document]
Comment (mattheworiordan) [reply spec/acceptance/realtime/auth_spec.rb]: Ok, so should we not have one test that ensures it reauths correctly without a disconnection?

File: spec/acceptance/realtime/auth_spec.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (funkyboy) [reply spec/acceptance/realtime/auth_spec.rb]: Oh now I understand what you mean :) see https://github.com/ably/ably-ruby/pull/137/commits/ffd1e18a2bcc309fe2320c9d5a6fc54f4af8c464

File: spec/acceptance/realtime/auth_spec.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (funkyboy) [reply spec/acceptance/realtime/auth_spec.rb]: Sure https://github.com/ably/ably-ruby/pull/137/commits/e9cdf14d269cf4ccc472cc14a602cea7198e1b57

File: spec/acceptance/rest/auth_spec.rb (hunk 7 above)
Comment (funkyboy) [reply spec/acceptance/rest/auth_spec.rb]: See https://github.com/ably/ably-ruby/commit/ffd1e18a2bcc309fe2320c9d5a6fc54f4af8c464

**Comment by funkyboy** [comment]
@paddybyers @mattheworiordan @SimonWoolf ready for another look

File: spec/acceptance/realtime/auth_spec.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (paddybyers) [reply spec/acceptance/realtime/auth_spec.rb]: I don't see how that change addresses the comment - the implication was that the test should be moved to the rest suite, not updated to become a realtime test :)

File: spec/acceptance/realtime/auth_spec.rb [hunk 14]
```diff
@@ -1031,5 +1031,184 @@ def disconnect_transport(connection)
...
+      end
+
+      context 'when the client is initialized with ClientOptions and the token is a JWT token' do
+        let(:client_options) { { token: token.body, environment: environment, protocol: protocol } }
+
+        context 'when credentials are valid' do
+          let(:token) { Faraday.get "#{auth_url}?keyName=#{key_name}&keySecret=#{key_secret}" }
+
+          it 'pulls stats successfully' do
+            client.stats do |stats|
```
[the remark this answers is not in the document]
Comment (paddybyers) [reply spec/acceptance/realtime/auth_spec.rb]: Again, that wasn't really the idea

File: spec/acceptance/realtime/auth_spec.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (funkyboy) [reply spec/acceptance/realtime/auth_spec.rb]: @paddybyers  We already have a [bunch of similar tests for rest that use stats](https://github.com/ably/ably-ruby/pull/137/files#diff-5fadc8e78f92977d3be0214a87031986R1343) so I thought of transforming this into one that uses realtime APIs.

[the remark this answers is not in the document]
Comment (paddybyers) [reply spec/acceptance/realtime/auth_spec.rb]: Sure, I'm not suggesting changing it now.

**Comment by funkyboy** [comment]
@mattheworiordan If you don't have any further observations I'd merge this

**Comment by mattheworiordan** [comment]
Merge away

**Pull request merged**